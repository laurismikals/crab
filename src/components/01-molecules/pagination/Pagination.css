.pagination {
  float: left;
}
.pagination__items {
  list-style: none;
  display: flex;
  flex-wrap: wrap;
}
.pagination__item {
}
.pagination__link {
  float: left;
  height: 40px;
  min-width: 40px;
  line-height: 40px;
  text-align: center;
  padding-right: 10px;
  padding-left: 10px;
  border-style: solid;
  border-color: #aaaaaa;
  border-width: 1px 0 1px 1px;
  background-color: #ffffff;
  transition: color 0.3s, background-color 0.3s;
}
.pagination__item:last-child .pagination__link {
  border-width: 1px;
}
a.pagination__link:hover,
a.pagination__link:focus{
  background-color: #eeeeee;
}
a.pagination__link.active {
  color: #ffffff;
  background-color: #aaaaaa;
}
