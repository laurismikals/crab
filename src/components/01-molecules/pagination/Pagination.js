import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/fp/noop';

import './Pagination.css';

const numberRangeArray = start => end =>
  new Array(end - start).fill().map((d, i) => i + start);

export class PaginationItem extends PureComponent {
  static propTypes = {
    isActive: PropTypes.bool,
    page: PropTypes.number.isRequired,
    onClick: PropTypes.func,
  };
  static defaultProps = {
    isActive: false,
    onClick: noop,
  };
  render() {
    const { page, isActive, onClick } = this.props;
    return (
      <li key={page} className="pagination__item">
        <a
          href=""
          tabIndex="1"
          className={`pagination__link ${isActive ? 'active' : ''}`}
          onClick={onClick}
        >
          {page}
        </a>
      </li>
    )
  }
}

export class Pagination extends PureComponent {
  static propTypes = {
    className: PropTypes.string,
    visiblePages: PropTypes.number,
    currentPage: PropTypes.number.isRequired,
    totalPages: PropTypes.number.isRequired,
    onClick: PropTypes.func,
  };
  static defaultProps = {
    className: '',
    visiblePages: 5,
    onClick: noop,
  };
  onClickHandler = (e, page) => {
    e.preventDefault();

    this.props.onClick(page);
  };
  render() {
    const {
      className,
      visiblePages,
      currentPage,
      totalPages,
    } = this.props;

    const wingSize = Math.floor(visiblePages / 2); // Number of pages visible before and after currentPage
    const separatorFill = 2; // Taking into account "..." separator cell. Separator shouldn't be visible in situation lik this - 1 ... 3, but it should be filled with a page like this - 1 2 3.

    const firstPageSeparationConditions =
      currentPage - visiblePages >= 0 &&
      totalPages - visiblePages - separatorFill > 0;
    // Checking if first page is separated from page range.
    const isFirstSeparate =
      currentPage >= visiblePages + wingSize || firstPageSeparationConditions;

    // Checking if last page is separated from page range
    const isLastSeparate =
      totalPages - separatorFill > currentPage + 1 + wingSize &&
      totalPages - (visiblePages + separatorFill) > 0;

    // Determining how many pages to add at the beginning of
    // the page range to make length of the range equal to visiblePages.
    const additionalPagesStart = !isLastSeparate ?
      currentPage >= totalPages - separatorFill ?
        separatorFill - (totalPages - currentPage) + 1 : 0
      :
      0;
    // Determining how many pages to add to the end of
    // the page range to make length of the range equal to visiblePages.
    const additionalPagesEnd = !isFirstSeparate ?
      wingSize < currentPage && currentPage < visiblePages ?
        currentPage + 1 + separatorFill : visiblePages
      :
      visiblePages;

    const rangeStart = !isFirstSeparate ? 0 : currentPage - wingSize - additionalPagesStart;
    const rangeEnd = !isLastSeparate ? totalPages : rangeStart + additionalPagesEnd;

    const pageArr = numberRangeArray(rangeStart)(rangeEnd);

    return (
      <nav className={`pagination ${className}`}>
        <ul className="pagination__items">
          {isFirstSeparate &&
            <PaginationItem
              page={1}
              onClick={(e) => this.onClickHandler(e, 0)}
            />
          }
          {isFirstSeparate &&
            <li className="pagination__item">
              <div className="pagination__link">
                ...
              </div>
            </li>
          }
          {pageArr.map((page) => {
            return (
              <PaginationItem
                key={page}
                page={page + 1}
                isActive={currentPage === page}
                onClick={(e) => this.onClickHandler(e, page)}
              />
            )
          })}
          {isLastSeparate &&
          <li className="pagination__item">
            <div className="pagination__link">
              ...
            </div>
          </li>
          }
          {isLastSeparate &&
          <PaginationItem
            page={totalPages}
            onClick={(e) => this.onClickHandler(e, totalPages - 1)}
          />
          }
        </ul>
      </nav>
    )
  }
}

export default Pagination;
