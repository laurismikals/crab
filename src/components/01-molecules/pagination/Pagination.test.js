import React from 'react';
import Enzyme, { shallow } from 'enzyme';

import { Pagination, PaginationItem } from './Pagination';

describe('Molecules: PaginationItem', () => {
  const minProps = {
    page: 1,
  };

  it('renders correctly', () => {
    const tree = shallow(
      <PaginationItem {...minProps} />
    );
    expect(tree).toMatchSnapshot();
  });
});

describe('Molecules: Pagination', () => {
  const minProps = {
    totalPages: 8,
  };

  it('renders correctly when current page is 1', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={0} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 2', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={1} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 3', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={2} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 4', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={3} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 5', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={4} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 6', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={5} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 7', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={6} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when current page is 8', () => {
    const tree = shallow(
      <Pagination {...minProps} currentPage={7} />
    );
    expect(tree).toMatchSnapshot();
  });
});
