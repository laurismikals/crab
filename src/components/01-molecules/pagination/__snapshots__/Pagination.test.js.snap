// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules: Pagination renders correctly when current page is 1 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={true}
      key="0"
      onClick={[Function]}
      page={1}
    />
    <PaginationItem
      isActive={false}
      key="1"
      onClick={[Function]}
      page={2}
    />
    <PaginationItem
      isActive={false}
      key="2"
      onClick={[Function]}
      page={3}
    />
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <li
      className="pagination__item"
    >
      <div
        className="pagination__link"
      >
        ...
      </div>
    </li>
    <PaginationItem
      isActive={false}
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 2 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      key="0"
      onClick={[Function]}
      page={1}
    />
    <PaginationItem
      isActive={true}
      key="1"
      onClick={[Function]}
      page={2}
    />
    <PaginationItem
      isActive={false}
      key="2"
      onClick={[Function]}
      page={3}
    />
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <li
      className="pagination__item"
    >
      <div
        className="pagination__link"
      >
        ...
      </div>
    </li>
    <PaginationItem
      isActive={false}
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 3 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      key="0"
      onClick={[Function]}
      page={1}
    />
    <PaginationItem
      isActive={false}
      key="1"
      onClick={[Function]}
      page={2}
    />
    <PaginationItem
      isActive={true}
      key="2"
      onClick={[Function]}
      page={3}
    />
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <li
      className="pagination__item"
    >
      <div
        className="pagination__link"
      >
        ...
      </div>
    </li>
    <PaginationItem
      isActive={false}
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 4 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      key="0"
      onClick={[Function]}
      page={1}
    />
    <PaginationItem
      isActive={false}
      key="1"
      onClick={[Function]}
      page={2}
    />
    <PaginationItem
      isActive={false}
      key="2"
      onClick={[Function]}
      page={3}
    />
    <PaginationItem
      isActive={true}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <PaginationItem
      isActive={false}
      key="5"
      onClick={[Function]}
      page={6}
    />
    <PaginationItem
      isActive={false}
      key="6"
      onClick={[Function]}
      page={7}
    />
    <PaginationItem
      isActive={false}
      key="7"
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 5 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      key="0"
      onClick={[Function]}
      page={1}
    />
    <PaginationItem
      isActive={false}
      key="1"
      onClick={[Function]}
      page={2}
    />
    <PaginationItem
      isActive={false}
      key="2"
      onClick={[Function]}
      page={3}
    />
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={true}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <PaginationItem
      isActive={false}
      key="5"
      onClick={[Function]}
      page={6}
    />
    <PaginationItem
      isActive={false}
      key="6"
      onClick={[Function]}
      page={7}
    />
    <PaginationItem
      isActive={false}
      key="7"
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 6 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      onClick={[Function]}
      page={1}
    />
    <li
      className="pagination__item"
    >
      <div
        className="pagination__link"
      >
        ...
      </div>
    </li>
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <PaginationItem
      isActive={true}
      key="5"
      onClick={[Function]}
      page={6}
    />
    <PaginationItem
      isActive={false}
      key="6"
      onClick={[Function]}
      page={7}
    />
    <PaginationItem
      isActive={false}
      key="7"
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 7 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      onClick={[Function]}
      page={1}
    />
    <li
      className="pagination__item"
    >
      <div
        className="pagination__link"
      >
        ...
      </div>
    </li>
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <PaginationItem
      isActive={false}
      key="5"
      onClick={[Function]}
      page={6}
    />
    <PaginationItem
      isActive={true}
      key="6"
      onClick={[Function]}
      page={7}
    />
    <PaginationItem
      isActive={false}
      key="7"
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: Pagination renders correctly when current page is 8 1`] = `
<nav
  className="pagination "
>
  <ul
    className="pagination__items"
  >
    <PaginationItem
      isActive={false}
      onClick={[Function]}
      page={1}
    />
    <li
      className="pagination__item"
    >
      <div
        className="pagination__link"
      >
        ...
      </div>
    </li>
    <PaginationItem
      isActive={false}
      key="3"
      onClick={[Function]}
      page={4}
    />
    <PaginationItem
      isActive={false}
      key="4"
      onClick={[Function]}
      page={5}
    />
    <PaginationItem
      isActive={false}
      key="5"
      onClick={[Function]}
      page={6}
    />
    <PaginationItem
      isActive={false}
      key="6"
      onClick={[Function]}
      page={7}
    />
    <PaginationItem
      isActive={true}
      key="7"
      onClick={[Function]}
      page={8}
    />
  </ul>
</nav>
`;

exports[`Molecules: PaginationItem renders correctly 1`] = `
<li
  className="pagination__item"
  key="1"
>
  <a
    className="pagination__link "
    href=""
    onClick={[Function]}
    tabIndex="1"
  >
    1
  </a>
</li>
`;
